keps/sig-network/3836-kube-proxy-improved-ingress-connectivity-reliability/README.md

```diff
@@ -0,0 +1,423 @@
+
+2. That Kube-proxy provides a mechanism for load balancers to do connection
+  draining for terminating Nodes. This is to be done by Kube-proxy by inspecting
+  a field on the Node object which indicates that the "node is
+  terminating/deleting" and when seen, starts failing the HC. When discussing
+  this scenario, the primary case this applied to was: downscaling by the
+  cluster autoscaler (CA). Given that the CA taints the Node which is to
+  be downscaled and deleted, with the taint: `ToBeDeletedByClusterAutoscaler` it
+  would seem most appropriate to use that here. It is unfortunate to have this
+  taint spread around the code base, but for now: no better indicator has been
+  thought of.
+
+3. This KEP does not attempt to align cloud providers for health checking
+   `eTP:Cluster` services. It recognizes cloud providers have valid reasons for
+   doing this differently depending on their implementations. However, it does
+   want to recommend ways for cloud providers to do health checking in a better
+   way on Kubernetes clusters. The KEP hence proposes that a document be added
+   to https://kubernetes.io/docs/concepts/ which can act as a formal guide and
+   be utilized as knowledge sharing with cloud providers.
+
+## Motivation
+
+The motivation for each change is:
+
+1. Not including the `healthz` state can cause Kube-proxy to indicate to a load
+  balancer that it should send traffic to a Node simply because the endpoint is
+  scheduled there, even though Kube-proxy might not be healthy and successfully
+  managed to write the rules required for actually being able to forward traffic
+  to the endpoint.
+
+2. Nodes used as intermediate nexthop for `eTP:Cluster` services would allow all
+  connections passing through the node while it is being terminated, to
+  gracefully shutdown. For `eTP:Local` workloads running on terminating nodes:
+  they get drained simultaneously with Kube-proxy today. This means that there's
+  no guarantee that connection draining will be enabled for them currently. That
+  might be the case if Kube-proxy is terminated before the endpoint, but there's
+  no guarantee of that. If Kube-proxy reacts to the
+  `ToBeDeletedByClusterAutoscaler` taint, then Kube-proxy can start prematurely
+  failing the HC and enable connection draining for these workloads.
+
+3. Cloud providers have very different ways of ascertaining if a load balancer
+   should target a specific node for `eTP:Cluster` services. We would like to
+   highlight the benefits of certain methods and pitfalls of others, in a formal
+   document, so that this is known. The hope is that this allows the information
+   to act as a source of knowledge for how to adapt their implementations to the
+   mechanics of a Kubernetes cluster.
+
+### Goals
+
+- Have kube-proxy accurately reflect its `healthz` state in its response to LB HCs
+  for `eTP:Local` services.
+
+- Offer a better capability of connection draining terminating Nodes, for load
```

Comment:
As we discussed on slack, I am concerned that changing the current /healthz behavior will have bad effect.

kube-proxy's /healthz serves two distinct purposes.  First, it indicates whether the kube-proxy is "healthy" (according to our internally defined metric of data-plane staleness).  Some people run kube-proxy as a Pod, and presumably point the `livenessProbe` at /healthz (though it looks like kubeadm doesn't set up liveness for kube-proxy).

Second, we point cloud LB healthchecks at /healthz to indicate "is this node a suitable 1st-hop backend (basically a router)  for this LB?".

We're talking about changing /healthz to consider both data-plane staleness AND some other thing(s) that would indicate this node is NOT a suitable 1st-hop backend (whether we use `unschedulable` or the `ToBeDeleted` taint or even node `deletionTimestamp`).

Changing /healthz will affect both of the above.  Do we really want kube-proxy to be restarted when the node becomes `unschedulable` ?  That doesn't seem right.  A node drain might take quite a while, and during that time the kube-proxy will fail liveness repeatedly and be restarted.

The advantage of making /healthz consider `unschedulable` (or whatever) is that clouds already point LBs at /healthz, so they will all get better "for free".

The alternative is to leave /healthz as it is and add /lbhealthz or something which is specifically geared for LB HCs.  The downside is that cloudf code needs to be updated and every existing LB needs to be updated to use it.

Another option could be to add a /livenessz URL with the "older" semantics and get people to use that for liveness.  That's strictly a break, but I can't honestly tell if anyone sets up liveness for kube-proxy.  kubeadm does not.  GKE does not.

Reply:
> Some people run kube-proxy as a Pod, and presumably point the `livenessProbe` at /healthz (though it looks like kubeadm doesn't set up liveness for kube-proxy).

we pushed back hard on that https://github.com/kubernetes/kubernetes/pull/75323 ;) , but is a possibility indeed

Reply:
IMO that argues for a calculated risk - using healthz for LBs is better all around and the risk of it breaking seems low.

How does this sound:

* refine /healthz as described in this KEP
* add /livez with the "historical" semantic
* (maybe) add /lb-healthz as an alias to /healthz with guidance to prefer this URL for clarity

?

Reply:
I am not sure it's worthwhile for us to go through all the hoops of doing points:

> * add /livez with the "historical" semantic
> * (maybe) add /lb-healthz as an alias to /healthz with guidance to prefer this URL for clarity

As you said: cloud-providers need to change their HC to benefit from this improvement (which many could argue, is in itself small/an optimization). And cloud providers won't change their HC per-service-proxy (i.e one URL for Kube-proxy, another for Cilium, etc). So for any of that to bare fruit we need to _also_ recommend and convince other service proxies to do the same....that feels so far away that it's practically impossible. 

Given that I feel we just do:

> * refine /healthz as described in this KEP     

and allow the outcome of calculated risks to surface when the KEP progresses through the different stages. If someone is configuring their Kube-proxy with a livenessProbe targeting `/healthz`, then we can decide to drop the feature then. This should, at worst, happen no later than Beta when the feature is on by default for everyone.

Reply:
If we think the risk is so significant, we can even extend the Beta phase?

Reply:
On top of it: this issue should quite quickly surface. Nodes go unschedulable/tainted all the time, and if the effect of that is a CrashLooping Kube-proxies: someone is likely to report that....I hope :/

Reply:
OK, that's a better argument I guess :) 

I will try to add that tomorrow morning.

Reply:
+1 for adding livez/ as an additional thing in this KEP